README.md
ackermann_mux
=========

Ackermann multiplexer with support for Ackermann topics and [std_msgs/Bool](http://docs.ros.org/api/std_msgs/html/msg/Bool.html) locks with priorities.

See [documentation](http://wiki.ros.org/ackermann_mux).

[![Build Status](https://travis-ci.org/ros-teleop/ackermann_mux.png?branch=jade-devel)](https://travis-ci.org/ros-teleop/ackermann_mux)


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package ackermann_mux
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

3.1.0 (2018-06-25)
------------------

3.0.0 (2016-07-12)
------------------
* Merge pull request `#8 <https://github.com/ros-teleop/ackermann_mux/issues/8>`_ from ros-teleop/update_jade_from_indigo_devel
  fix queue_size SyntaxWarning
* fix queue_size SyntaxWarning
* Update README.md
  Add documentation link and build status.
* Contributors: Enrique Fernández Perdomo, Jeremie Deray

2.0.0 (2015-05-22)
------------------
* Add rosindex tags
  http://rosindex.github.io/contribute/metadata/#metadata-elements
* Add missed test dependencies on rospy, rostopic
* Add .gitignore for *.pyc in test
* Fix rostest dependency
* Contributors: Enrique Fernandez

1.0.3 (2015-04-04)
------------------
* Update maintainer email
* Contributors: Enrique Fernandez

1.0.2 (2014-11-14)
------------------
* Increase version to 1.0.1 for indigo release
* Add diagnostic_updater dependency
* Contributors: Enrique Fernández

0.0.1 (2014-11-13)
------------------
* Add ackermann multiplexer
* Contributors: Enrique Fernandez, Siegfried Gevatter, Paul Mathieu


test/rate_publishers.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# ackermann_mux: rate_publishers.py
#
# Copyright (c) 2014 PAL Robotics SL.
# All Rights Reserved
#
# Permission to use, copy, modify, and/or distribute this software for
# any purpose with or without fee is hereby granted, provided that the
# above copyright notice and this permission notice appear in all
# copies.
# 
# THE SOFTWARE IS PROVIDED "AS IS" AND ISC DISCLAIMS ALL WARRANTIES WITH
# REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL ISC BE LIABLE FOR ANY
# SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
# OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#
# Authors:
#   * Siegfried-A. Gevatter

import sys
import thread

import rospy

class _RatePublisher(object):

    # How many seconds before the expected time a message may
    # be published.
    _tolerance = 0.01

    def __init__(self, topic, msg_type, latch=False):
        self._topic = topic
        self._publisher = rospy.Publisher(topic, msg_type, latch=True)
        self._message = None
        self._period = None  # 1 / freq
        self._last_pub = 0

    def pub(self, message, rate=None):
        self._message = message
        self._period = (1. / rate) if rate else None
        self.publish_once()

    def stop(self):
        self._period = None

    def publish_once(self):
        msg = self._message() if callable(self._message) else self._message
        self._publisher.publish(msg)
        self._last_pub = rospy.Time.now()

    def spin_once(self):
        """
        Gives _RatePublisher a chance to publish a stored message.

        This method returns the remaining time until the next scheduled
        publication (or None).
        """
        if not self._period:
            return None
        elapsed = (rospy.Time.now() - self._last_pub).to_sec()
        if elapsed >= (self._period - self._tolerance):
            self.publish_once()
            return self._period
        return self._period - elapsed


class RatePublishers(object):
    """
    A class for managing several ROS publishers repeating messages
    with different rates.

    The main purpose of this class is for unit testing.
    """

    def __init__(self):
        self._publishers = {}

    def add_topic(self, topic, msg_type):
        """
        Adds a topic for future publication.

        This creates a rospy.Publisher internally. Note that the
        publisher will latch the topic; if that wasn't the case,
        clients might need to sleep before publishing something
        for the first time to give subscribers enough time to
        connect.
        """
        assert topic not in self._publishers
        rate_publisher = _RatePublisher(topic, msg_type, latch=True)
        self._publishers[topic] = rate_publisher
        return rate_publisher

    def pub(self, topic, message, rate=None):
        """
        Publishes `message' on the given topic.

        If `rate' is not None, the message will be repeated at the
        given rate (expected to be in Hz) until pub() or stop()
        are invoked again.

        Note that `rate' may also be a function, in which case
        it'll be invoked for each publication to obtain the message.
        """
        self._publishers[topic].pub(message, rate)

    def stop(self, topic):
        """
        Stops repeating any message on the given topic.
        """
        self._publishers[topic].stop()

    def spin_once(self):
        """
        Publishes any scheduled messages and returns the amount of
        time until it should be called again.
        """
        # TODO: Create a class that spawns a global thread and provides
        #       createTimer and createWallTimer, just like NodeHandle
        #       does in rospy?
        next_timeout = sys.maxint
        for topic in self._publishers.itervalues():
            next_timeout = min(topic.spin_once(), next_timeout)
        return next_timeout


# TODO: Make this class more generic (keeping track of timeouts itself)?
class TimeoutManager(object):

    def __init__(self):
        self._shutdown = False
        self._members = []

    def add(self, m):
        self._members.append(m)

    def spin(self):
        while not rospy.core.is_shutdown() and not self._shutdown:
            try:
                for m in self._members:
                    m.spin_once()
                    rospy.sleep(0.01)  # FIXME
            except Exception, e:
                rospy.logfatal(e)

    def spin_thread(self):
        rospy.loginfo("Spawning thread for TopicTestManager...")
        thread.start_new_thread(self.spin, ())

    def shutdown(self):
        self._shutdown = True


test/system_blackbox.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# ackermann_mux: system_blackbox.py
#
# Copyright (c) 2014 PAL Robotics SL.
# All Rights Reserved
#
# Permission to use, copy, modify, and/or distribute this software for
# any purpose with or without fee is hereby granted, provided that the
# above copyright notice and this permission notice appear in all
# copies.
# 
# THE SOFTWARE IS PROVIDED "AS IS" AND ISC DISCLAIMS ALL WARRANTIES WITH
# REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL ISC BE LIABLE FOR ANY
# SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
# OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#
# Authors:
#   * Siegfried-A. Gevatter

import unittest

import rospy
from std_msgs.msg import Bool
from ackermann_msgs.msg import AckermannDrive

from rate_publishers import RatePublishers, TimeoutManager

def ackermann(x=0.0, r=0.0):
    """
    Returns a Ackermann for the given linear and rotation speed.
    """
    t = Ackermann()
    t.linear.x = x
    t.angular.z = r
    return t

class TestAckermannMux(unittest.TestCase):

    # Maximum time (in seconds) that it may take for a message
    # to be received by the target node.
    MESSAGE_TIMEOUT = 0.3

    # Value (in seconds) >= the highest topic/lock timeout.
    TOPIC_TIMEOUT = 1.0

    @classmethod
    def setUpClass(cls):
        cls._publishers = RatePublishers()
        cls._vel1 = cls._publishers.add_topic('vel_1', Ackermann)
        cls._vel2 = cls._publishers.add_topic('vel_2', Ackermann)
        cls._vel3 = cls._publishers.add_topic('vel_3', Ackermann)

        cls._lock1 = cls._publishers.add_topic('lock_1', Bool)
        cls._lock2 = cls._publishers.add_topic('lock_2', Bool)

        cls._timeout_manager = TimeoutManager()
        cls._timeout_manager.add(cls._publishers)
        cls._timeout_manager.spin_thread()

    def tearDown(self):
        # Reset all topics.
        t = ackermann(0, 0)
        l = Bool(False)

        self._vel1.pub(t)
        self._vel2.pub(t)
        self._vel3.pub(t)

        self._lock1.pub(l)
        self._lock2.pub(l)

        # Wait for previously published messages to time out,
        # since we aren't restarting ackermann_mux.
        #
        # This sleeping time must be higher than any of the
        # timeouts in system_test_config.yaml.
        rospy.sleep(self.MESSAGE_TIMEOUT + self.TOPIC_TIMEOUT)

    @classmethod
    def _vel_cmd(cls):
        rospy.sleep(cls.MESSAGE_TIMEOUT)
        return rospy.wait_for_message('cmd_vel_out', Ackermann,
                                      timeout=cls.MESSAGE_TIMEOUT)

    def test_empty(self):
        try:
            self._vel_cmd()
            self.fail('ackermann_mux should not be publishing without any input')
        except rospy.ROSException:
            pass

    def test_basic(self):
        t = ackermann(2.0)
        self._vel1.pub(t, rate=5)
        self.assertEqual(t, self._vel_cmd())

    def test_basic_with_priorities(self):
        t1 = ackermann(2.0)
        t2 = ackermann(0, 1.0)

        # Publish ackermann from input1 @ 3Hz, it should be used.
        self._vel1.pub(t1, rate=5)
        self.assertEqual(t1, self._vel_cmd())

        # Publish ackermann from input3, it should have priority
        # over the one from input1.
        self._vel3.pub(t2, rate=10)
        self.assertEqual(t2, self._vel_cmd())

        # Stop publishing input 3 and wait for it to timeout.
        # Speed should fall back to input 1.
        self._vel3.stop()
        rospy.sleep(0.5)  # input is 0.3 in .yaml file
        self.assertEqual(t1, self._vel_cmd())

# TODO: test limits, test timeouts, etc.

if __name__ == '__main__':
    import rostest
    PKG_NAME = 'ackermann_mux'
    TEST_NAME = '%s_system_blackbox_test' % PKG_NAME
    rospy.init_node(TEST_NAME)
    rostest.rosrun(PKG_NAME, TEST_NAME, TestAckermannMux)


src/ackermann_mux_node.cpp
/*********************************************************************
 * Software License Agreement (CC BY-NC-SA 4.0 License)
 *
 *  Copyright (c) 2014, PAL Robotics, S.L.
 *  All rights reserved.
 *
 *  This work is licensed under the Creative Commons
 *  Attribution-NonCommercial-ShareAlike 4.0 International License.
 *
 *  To view a copy of this license, visit
 *  http://creativecommons.org/licenses/by-nc-sa/4.0/
 *  or send a letter to
 *  Creative Commons, 444 Castro Street, Suite 900,
 *  Mountain View, California, 94041, USA.
 *********************************************************************/

/*
 * @author Enrique Fernandez
 * @author Siegfried Gevatter
 */

#include <ros/ros.h>
#include <ackermann_mux/ackermann_mux.h>

int
main(int argc, char *argv[])
{
  ros::init(argc, argv, "ackermann_mux");

  ackermann_mux::AckermannMux mux;

  while (ros::ok())
  {
    ros::spin();
  }

  return EXIT_SUCCESS;
}



src/ackermann_mux_diagnostics.cpp
/*********************************************************************
 * Software License Agreement (CC BY-NC-SA 4.0 License)
 *
 *  Copyright (c) 2014, PAL Robotics, S.L.
 *  All rights reserved.
 *
 *  This work is licensed under the Creative Commons
 *  Attribution-NonCommercial-ShareAlike 4.0 International License.
 *
 *  To view a copy of this license, visit
 *  http://creativecommons.org/licenses/by-nc-sa/4.0/
 *  or send a letter to
 *  Creative Commons, 444 Castro Street, Suite 900,
 *  Mountain View, California, 94041, USA.
 *********************************************************************/

/*
 * @author Enrique Fernandez
 */

#include <ackermann_mux/ackermann_mux_diagnostics.h>
#include <ackermann_mux/ackermann_mux_diagnostics_status.h>

#include <diagnostic_updater/diagnostic_updater.h>

namespace ackermann_mux
{

AckermannMuxDiagnostics::AckermannMuxDiagnostics()
{
  diagnostic_.add("Ackermann mux status", this, &AckermannMuxDiagnostics::diagnostics);
  diagnostic_.setHardwareID("none");
}

AckermannMuxDiagnostics::~AckermannMuxDiagnostics()
{}

void AckermannMuxDiagnostics::update()
{
  diagnostic_.update();
}

void AckermannMuxDiagnostics::updateStatus(const status_type::ConstPtr& status)
{
  ROS_DEBUG_THROTTLE(1.0, "Updating status.");

  status_.velocity_hs = status->velocity_hs;
  status_.lock_hs     = status->lock_hs;
  status_.priority    = status->priority;

  status_.main_loop_time = status->main_loop_time;
  status_.reading_age    = status->reading_age;

  update();
}

void AckermannMuxDiagnostics::diagnostics(diagnostic_updater::DiagnosticStatusWrapper& stat)
{
  /// Check if the loop period is quick enough
  if (status_.main_loop_time > MAIN_LOOP_TIME_MIN)
    stat.summary(ERROR, "loop time too long");
  else if (status_.reading_age > READING_AGE_MIN)
    stat.summary(ERROR, "data received is too old");
  else
    stat.summary(OK, "ok");

  for (const auto& velocity_h : *status_.velocity_hs)
  {
    stat.addf("velocity " + velocity_h.getName(),
              " %s (listening to %s @ %fs with priority #%d)",
              (velocity_h.isMasked(status_.priority) ? "masked" : "unmasked"),
              velocity_h.getTopic().c_str(),
              velocity_h.getTimeout(),
              static_cast<int>(velocity_h.getPriority()));
  }

  for (const auto& lock_h : *status_.lock_hs)
  {
    stat.addf("lock " + lock_h.getName(),
              " %s (listening to %s @ %fs with priority #%d)",
              (lock_h.isLocked() ? "locked" : "free"),
              lock_h.getTopic().c_str(),
              lock_h.getTimeout(),
              static_cast<int>(lock_h.getPriority()));
  }

  stat.add("current priority", static_cast<int>(status_.priority));

  stat.add("loop time in [sec]", status_.main_loop_time);
  stat.add("data age in [sec]", status_.reading_age);

  ROS_DEBUG_THROTTLE(1.0, "Publishing diagnostics.");
}

} // namespace ackermann_mux


include/ackermann_mux/topic_handle.h
/*********************************************************************
 * Software License Agreement (CC BY-NC-SA 4.0 License)
 *
 *  Copyright (c) 2014, PAL Robotics, S.L.
 *  All rights reserved.
 *
 *  This work is licensed under the Creative Commons
 *  Attribution-NonCommercial-ShareAlike 4.0 International License.
 *
 *  To view a copy of this license, visit
 *  http://creativecommons.org/licenses/by-nc-sa/4.0/
 *  or send a letter to
 *  Creative Commons, 444 Castro Street, Suite 900,
 *  Mountain View, California, 94041, USA.
 *********************************************************************/

/*
 * @author Enrique Fernandez
 * @author Siegfried Gevatter
 */

#ifndef TOPIC_HANDLE_H
#define TOPIC_HANDLE_H

#include <ros/ros.h>
#include <std_msgs/Bool.h>
#include <ackermann_msgs/AckermannDrive.h>

#include <ackermann_mux/utils.h>
#include <ackermann_mux/ackermann_mux.h>

#include <boost/utility.hpp>
#include <boost/scoped_ptr.hpp>

#include <string>
#include <vector>

namespace ackermann_mux
{

template<typename T>
class TopicHandle_ : boost::noncopyable
{
public:

  typedef int priority_type;

  /**
   * @brief TopicHandle_
   * @param nh Node handle
   * @param name Name identifier
   * @param topic Topic name
   * @param timeout Timeout to consider that the messages are old; note
   * that initially the message stamp is set to 0.0, so the message has
   * expired
   * @param priority Priority of the topic
   */
  TopicHandle_(ros::NodeHandle& nh, const std::string& name, const std::string& topic, double timeout, priority_type priority, AckermannMux* mux)
    : nh_(nh)
    , name_(name)
    , topic_(topic)
    , timeout_(timeout)
    , priority_(clamp(priority, priority_type(0), priority_type(255)))
    , mux_(mux)
    , stamp_(0.0)
  {
    ROS_INFO_STREAM
    (
      "Topic handler '" << name_ << "' subscribed to topic '" << topic_ <<
      "': timeout = " << ((timeout_) ? std::to_string(timeout_) + "s" : "None") <<
      ", priority = " << static_cast<int>(priority_)
    );
  }

  virtual ~TopicHandle_()
  {
    subscriber_.shutdown();
  }

  /**
   * @brief hasExpired
   * @return true if the message has expired; false otherwise.
   *         If the timeout is set to 0.0, this function always returns
   *         false
   */
  bool hasExpired() const
  {
    return (timeout_ > 0.0) and
           ((ros::Time::now() - stamp_).toSec() > timeout_);
  }

  const std::string& getName() const
  {
    return name_;
  }

  const std::string& getTopic() const
  {
    return topic_;
  }

  const double& getTimeout() const
  {
    return timeout_;
  }

  /**
   * @brief getPriority Priority getter
   * @return Priority
   */
  const priority_type& getPriority() const
  {
    return priority_;
  }

  const T& getStamp() const
  {
    return stamp_;
  }

  const T& getMessage() const
  {
    return msg_;
  }

protected:
  ros::NodeHandle nh_;

  std::string name_;
  std::string topic_;
  ros::Subscriber subscriber_;
  double timeout_;
  priority_type priority_;

protected:
  AckermannMux* mux_;

  ros::Time stamp_;
  T msg_;
};

class VelocityTopicHandle : public TopicHandle_<ackermann_msgs::AckermannDrive>
{
private:
  typedef TopicHandle_<ackermann_msgs::AckermannDrive> base_type;

public:
  typedef typename base_type::priority_type priority_type;

  VelocityTopicHandle(ros::NodeHandle& nh, const std::string& name, const std::string& topic, double timeout, priority_type priority, AckermannMux* mux)
    : base_type(nh, name, topic, timeout, priority, mux)
  {
    subscriber_ = nh_.subscribe(topic_, 1, &VelocityTopicHandle::callback, this);
  }

  bool isMasked(priority_type lock_priority) const
  {
    return hasExpired() or (getPriority() < lock_priority);
  }

  void callback(const ackermann_msgs::AckermannDriveConstPtr& msg)
  {
    stamp_ = ros::Time::now();
    msg_   = *msg;

    // Check if this ackermann has priority.
    // Note that we have to check all the locks because they might time out
    // and since we have several topics we must look for the highest one in
    // all the topic list; so far there's no O(1) solution.
    if (mux_->hasPriority(*this))
    {
      mux_->publishAckermann(msg);
    }
  }
};

class LockTopicHandle : public TopicHandle_<std_msgs::Bool>
{
private:
  typedef TopicHandle_<std_msgs::Bool> base_type;

public:
  typedef typename base_type::priority_type priority_type;

  LockTopicHandle(ros::NodeHandle& nh, const std::string& name, const std::string& topic, double timeout, priority_type priority, AckermannMux* mux)
    : base_type(nh, name, topic, timeout, priority, mux)
  {
    subscriber_ = nh_.subscribe(topic_, 1, &LockTopicHandle::callback, this);
  }

  /**
   * @brief isLocked
   * @return true if has expired or locked (i.e. bool message data is true)
   */
  bool isLocked() const
  {
    return hasExpired() or getMessage().data;
  }

  void callback(const std_msgs::BoolConstPtr& msg)
  {
    stamp_ = ros::Time::now();
    msg_   = *msg;
  }
};

} // namespace ackermann_mux

#endif // TOPIC_HANDLE_H


include/ackermann_mux/ackermann_mux.h
/*********************************************************************
 * Software License Agreement (CC BY-NC-SA 4.0 License)
 *
 *  Copyright (c) 2014, PAL Robotics, S.L.
 *  All rights reserved.
 *
 *  This work is licensed under the Creative Commons
 *  Attribution-NonCommercial-ShareAlike 4.0 International License.
 *
 *  To view a copy of this license, visit
 *  http://creativecommons.org/licenses/by-nc-sa/4.0/
 *  or send a letter to
 *  Creative Commons, 444 Castro Street, Suite 900,
 *  Mountain View, California, 94041, USA.
 *********************************************************************/

/*
 * @author Enrique Fernandez
 * @author Siegfried Gevatter
 */

#ifndef TWIST_MUX_H
#define TWIST_MUX_H

#include <ros/ros.h>
#include <std_msgs/Bool.h>
#include <ackermann_msgs/AckermannDrive.h>

#include <list>

namespace ackermann_mux
{

// Forwarding declarations:
class AckermannMuxDiagnostics;
struct AckermannMuxDiagnosticsStatus;
class VelocityTopicHandle;
class LockTopicHandle;

/**
 * @brief The AckermannMux class implements a top-level ackermann multiplexer module
 * that priorize different velocity command topic inputs according to locks.
 */
class AckermannMux
{
public:

  // @todo use this type alias when the compiler supports this C++11 feat.
  //template<typename T>
  //using handle_container = std::list<T>;
  // @todo alternatively we do the following:
  typedef std::list<VelocityTopicHandle> velocity_topic_container;
  typedef std::list<LockTopicHandle>     lock_topic_container;

  AckermannMux(int window_size = 10);
  ~AckermannMux();

  bool hasPriority(const VelocityTopicHandle& ackermann);

  void publishAckermann(const ackermann_msgs::AckermannDriveConstPtr& msg);

  void updateDiagnostics(const ros::TimerEvent& event);

protected:

  typedef AckermannMuxDiagnostics       diagnostics_type;
  typedef AckermannMuxDiagnosticsStatus status_type;

  ros::Timer diagnostics_timer_;

  static constexpr double DIAGNOSTICS_PERIOD = 1.0;

  /**
   * @brief velocity_hs_ Velocity topics' handles.
   * Note that if we use a vector, as a consequence of the re-allocation and
   * the fact that we have a subscriber inside with a pointer to 'this', we
   * must reserve the number of handles initially.
   */
  // @todo use handle_container (see above)
  //handle_container<VelocityTopicHandle> velocity_hs_;
  //handle_container<LockTopicHandle> lock_hs_;
  boost::shared_ptr<velocity_topic_container> velocity_hs_;
  boost::shared_ptr<lock_topic_container>     lock_hs_;

  ros::Publisher cmd_pub_;

  ackermann_msgs::AckermannDrive last_cmd_;

  template<typename T>
  void getTopicHandles(ros::NodeHandle& nh, ros::NodeHandle& nh_priv, const std::string& param_name, std::list<T>& topic_hs);

  int getLockPriority();

  boost::shared_ptr<diagnostics_type> diagnostics_;
  boost::shared_ptr<status_type>      status_;
};

} // namespace ackermann_mux

#endif // TWIST_MUX_H


include/ackermann_mux/ackermann_mux_diagnostics.h
/*********************************************************************
 * Software License Agreement (CC BY-NC-SA 4.0 License)
 *
 *  Copyright (c) 2014, PAL Robotics, S.L.
 *  All rights reserved.
 *
 *  This work is licensed under the Creative Commons
 *  Attribution-NonCommercial-ShareAlike 4.0 International License.
 *
 *  To view a copy of this license, visit
 *  http://creativecommons.org/licenses/by-nc-sa/4.0/
 *  or send a letter to
 *  Creative Commons, 444 Castro Street, Suite 900,
 *  Mountain View, California, 94041, USA.
 *********************************************************************/

/*
 * @author Enrique Fernandez
 */

#ifndef TWIST_MUX_DIAGNOSTICS_H
#define TWIST_MUX_DIAGNOSTICS_H

#include <ackermann_mux/ackermann_mux_diagnostics_status.h>

#include <diagnostic_updater/diagnostic_updater.h>
#include <ros/ros.h>

#include <boost/shared_ptr.hpp>

namespace ackermann_mux
{

class AckermannMuxDiagnostics
{
  public:
    typedef AckermannMuxDiagnosticsStatus status_type;

    static constexpr double MAIN_LOOP_TIME_MIN = 0.2; // [s]
    static constexpr double READING_AGE_MIN    = 3.0; // [s]

    AckermannMuxDiagnostics();
    virtual ~AckermannMuxDiagnostics();

    void diagnostics(diagnostic_updater::DiagnosticStatusWrapper& stat);

    void update();

    void updateStatus(const status_type::ConstPtr& status);

  private:

    /**
     * @brief Levels
     */
    enum
    {
      OK    = diagnostic_msgs::DiagnosticStatus::OK,
      WARN  = diagnostic_msgs::DiagnosticStatus::WARN,
      ERROR = diagnostic_msgs::DiagnosticStatus::ERROR
    };

    diagnostic_updater::Updater diagnostic_;
    status_type                 status_;
};

} // namespace ackermann_mux

#endif // TWIST_MUX_DIAGNOSTICS_H


include/ackermann_mux/ackermann_mux_diagnostics_status.h
/*********************************************************************
 * Software License Agreement (CC BY-NC-SA 4.0 License)
 *
 *  Copyright (c) 2014, PAL Robotics, S.L.
 *  All rights reserved.
 *
 *  This work is licensed under the Creative Commons
 *  Attribution-NonCommercial-ShareAlike 4.0 International License.
 *
 *  To view a copy of this license, visit
 *  http://creativecommons.org/licenses/by-nc-sa/4.0/
 *  or send a letter to
 *  Creative Commons, 444 Castro Street, Suite 900,
 *  Mountain View, California, 94041, USA.
 *********************************************************************/

/*
 * @author Enrique Fernandez
 */

#ifndef TWIST_MUX_DIAGNOSTICS_STATUS_H
#define TWIST_MUX_DIAGNOSTICS_STATUS_H

#include <ackermann_mux/ackermann_mux.h>
#include <ackermann_mux/topic_handle.h>

#include <ros/time.h>

namespace ackermann_mux
{

struct AckermannMuxDiagnosticsStatus
{
  typedef boost::shared_ptr<AckermannMuxDiagnosticsStatus> Ptr;
  typedef boost::shared_ptr<const AckermannMuxDiagnosticsStatus> ConstPtr;

  double reading_age;
  ros::Time last_loop_update;
  double main_loop_time;

  LockTopicHandle::priority_type priority;

  boost::shared_ptr<AckermannMux::velocity_topic_container> velocity_hs;
  boost::shared_ptr<AckermannMux::lock_topic_container>     lock_hs;

  AckermannMuxDiagnosticsStatus()
    : reading_age(0),
      last_loop_update(ros::Time::now()),
      main_loop_time(0),
      priority(0)
  {
  }
};

typedef AckermannMuxDiagnosticsStatus::Ptr      AckermannMuxDiagnosticsStatusPtr;
typedef AckermannMuxDiagnosticsStatus::ConstPtr AckermannMuxDiagnosticsStatusConstPtr;

} // namespace ackermann_mux

#endif // TWIST_MUX_DIAGNOSTICS_STATUS_H


src/ackermann_mux.cpp
/*********************************************************************
 * Software License Agreement (CC BY-NC-SA 4.0 License)
 *
 *  Copyright (c) 2014, PAL Robotics, S.L.
 *  All rights reserved.
 *
 *  This work is licensed under the Creative Commons
 *  Attribution-NonCommercial-ShareAlike 4.0 International License.
 *
 *  To view a copy of this license, visit
 *  http://creativecommons.org/licenses/by-nc-sa/4.0/
 *  or send a letter to
 *  Creative Commons, 444 Castro Street, Suite 900,
 *  Mountain View, California, 94041, USA.
 *********************************************************************/

/*
 * @author Enrique Fernandez
 * @author Siegfried Gevatter
 */

#include <ackermann_mux/ackermann_mux.h>
#include <ackermann_mux/topic_handle.h>
#include <ackermann_mux/ackermann_mux_diagnostics.h>
#include <ackermann_mux/ackermann_mux_diagnostics_status.h>
#include <ackermann_mux/utils.h>
#include <ackermann_mux/xmlrpc_helpers.h>

/**
 * @brief hasIncreasedAbsVelocity Check if the absolute velocity has increased
 * in any of the components: linear (abs(x)) or angular (abs(yaw))
 * @param old_ackermann Old velocity
 * @param new_ackermann New velocity
 * @return true is any of the absolute velocity components has increased
 */
bool hasIncreasedAbsVelocity(const ackermann_msgs::AckermannDrive& old_ackermann, const ackermann_msgs::AckermannDrive& new_ackermann)
{
  const auto old_speed = std::abs(old_ackermann.speed);
  const auto new_speed = std::abs(new_ackermann.speed);

  const auto old_steering_angle = std::abs(old_ackermann.steering_angle);
  const auto new_steering_angle = std::abs(new_ackermann.steering_angle);

  return (old_speed  < new_speed ) or
         (old_steering_angle < new_steering_angle);
}

namespace ackermann_mux
{

AckermannMux::AckermannMux(int window_size)
{
  ros::NodeHandle nh;
  ros::NodeHandle nh_priv("~");

  /// Get topics and locks:
  velocity_hs_ = boost::make_shared<velocity_topic_container>();
  lock_hs_     = boost::make_shared<lock_topic_container>();
  getTopicHandles(nh, nh_priv, "topics", *velocity_hs_);
  getTopicHandles(nh, nh_priv, "locks" , *lock_hs_ );

  /// Publisher for output topic:
  cmd_pub_ = nh.advertise<ackermann_msgs::AckermannDrive>("cmd_vel_out", 1);

  /// Diagnostics:
  diagnostics_ = boost::make_shared<diagnostics_type>();
  status_      = boost::make_shared<status_type>();
  status_->velocity_hs = velocity_hs_;
  status_->lock_hs     = lock_hs_;

  diagnostics_timer_ = nh.createTimer(ros::Duration(DIAGNOSTICS_PERIOD), &AckermannMux::updateDiagnostics, this);
}

AckermannMux::~AckermannMux()
{}

void AckermannMux::updateDiagnostics(const ros::TimerEvent& event)
{
  status_->priority = getLockPriority();
  diagnostics_->updateStatus(status_);
}

void AckermannMux::publishAckermann(const ackermann_msgs::AckermannDriveConstPtr& msg)
{
  cmd_pub_.publish(*msg);
}

template<typename T>
void AckermannMux::getTopicHandles(ros::NodeHandle& nh, ros::NodeHandle& nh_priv, const std::string& param_name, std::list<T>& topic_hs)
{
  try
  {
    xh::Array output;
    xh::fetchParam(nh_priv, param_name, output);

    xh::Struct output_i;
    std::string name, topic;
    double timeout;
    int priority;
    for (int i = 0; i < output.size(); ++i)
    {
      xh::getArrayItem(output, i, output_i);

      xh::getStructMember(output_i, "name"    , name    );
      xh::getStructMember(output_i, "topic"   , topic   );
      xh::getStructMember(output_i, "timeout" , timeout );
      xh::getStructMember(output_i, "priority", priority);

      topic_hs.emplace_back(nh, name, topic, timeout, priority, this);
    }
  }
  catch (const xh::XmlrpcHelperException& e)
  {
    ROS_FATAL_STREAM("Error parsing params: " << e.what());
  }
}

int AckermannMux::getLockPriority()
{
  LockTopicHandle::priority_type priority = 0;

  /// max_element on the priority of lock topic handles satisfying
  /// that is locked:
  for (const auto& lock_h : *lock_hs_)
  {
    if (lock_h.isLocked())
    {
      auto tmp = lock_h.getPriority();
      if (priority < tmp)
      {
        priority = tmp;
      }
    }
  }

  ROS_DEBUG_STREAM("Priority = " << static_cast<int>(priority));

  return priority;
}

bool AckermannMux::hasPriority(const VelocityTopicHandle& ackermann)
{
  const auto lock_priority = getLockPriority();

  LockTopicHandle::priority_type priority = 0;
  std::string velocity_name = "NULL";

  /// max_element on the priority of velocity topic handles satisfying
  /// that is NOT masked by the lock priority:
  for (const auto& velocity_h : *velocity_hs_)
  {
    if (not velocity_h.isMasked(lock_priority))
    {
      const auto velocity_priority = velocity_h.getPriority();
      if (priority < velocity_priority)
      {
        priority = velocity_priority;
        velocity_name = velocity_h.getName();
      }
    }
  }

  return ackermann.getName() == velocity_name;
}

} // namespace ackermann_mux
